### ISSUE TITLE
Support the equivalent of $elemMatch

### ISSUE THREAD

**Issue by jyemin** [NathanQingyangXu/jpa-mongodb-mapping#50]
MongoDB supports a powerful query operator, [elemMatch](https://www.mongodb.com/docs/manual/reference/operator/query/elemMatch/), that lets applications perform "joins" on embedded arrays of documents.  Can we support this with existing functions in Hibernate, or if not can we define a custom function to do so?  For example, given this entity and struct:

```
    @Entity(name = "Movie")
    class Movie {
        @Id
        int id;
        List<TagsByAuthor> tagsByAuthorList;
    }

    @Embeddable
    @Struct(name = "TagsByAuthor")
    class TagsByAuthor {
       String author;
       List<String> tags;
    }
```

If I want to find all movies with a tagsByAuthorList that contains a document with a given author and a given tag, I could do something like this in MQL

```
{tagsByAuthorList : {$elemMatch : {author : "jeff", tags : "funny"}}}
```

How could this be expressed in HQL?  Perhaps something like:

```
from Movie where 
array_element_matches(tagsByAuthorList, author = 'jeff' and array_contains(tags, 'funny')
```

?

But could the HQL parser interpret that correctly?  It would have to know that _author_ and _tags_ refer to fields of _TagsByAuthor_, but how would it be able to determine that?

**Comment by NathanQingyangXu** [comment]
I highly doubt Hibernate supports this (I checked the few STRUCT related PRs created by Christian Beikov and found none of them contains even basic query testing cases; the only one provided is based on entity id). For now it is more realistic to rely on native query to go about such advanced usage.

**Comment by jyemin** [comment]
The problem is that for a document database like MongoDB this is not advanced usage at all.

**Comment by jyemin** [comment]
I asked about it on Zulip: https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev/topic/Custom.20functions.

**Comment by NathanQingyangXu** [comment]
From my understanding, array function currently is very basic. But to my
knowledge, SQL function doesn't support a predicate or another function as
parameter. Without such HQL support, it might be hard to translate as
expected.

**Comment by jyemin** [comment]
Got it.  This seems like a serious issue then.

**Comment by NathanQingyangXu** [comment]
SQL is procedure stuff. I would be really amazed if it is modernized to embrace functional programming. How many years went by before Java introduces Lambda finally. SQL's backward compatibility burden would be so much more.
Here is a discussion on Stackoverflow on this: https://stackoverflow.com/questions/1153397/besides-a-declarative-language-is-sql-a-functional-language

**Comment by NathanQingyangXu** [comment]
As Christian Beikov answered on Zulip, Hibernate v7 will introduce a new `unnest` function to solve the issue.

**Comment by jyemin** [comment]
`unnest`, rather. :). 

But I like this suggestion better if we can make it work:

```
from Movie m
join m.tagsByAuthorList t
where t.author = :author
and t.tag = :tag
``

**Comment by NathanQingyangXu** [comment]
But it requires a join or another collection, right? That would lose the attraction of Mongo's embedded BSon array or Bson document, right?

**Comment by jyemin** [comment]
No, not at all, iirc.  It would be specified as a join in HQL, but rendered as `$elemMatch` in our dialect (I'm not sure _how_ we would do that but I think it would be possible)

**Comment by NathanQingyangXu** [comment]
got it. that being said, HQL could not be tweaked easily without ripple effect so we should be extremely cautious about that. For one thing, it might not align with HQL grammer so runtime exception might be thrown (it could be validated during compile time if we use Jarkarta Data Repository tool as per https://docs.jboss.org/hibernate/orm/6.6/repositories/html_single/Hibernate_Data_Repositories.html).

**Comment by NathanQingyangXu** [comment]
But if it won't break HQL grammer, it might be a creative way to solve the issue, hack as it is.

**Comment by NathanQingyangXu** [comment]
I created a testing case as a placeholder (https://github.com/NathanQingyangXu/jpa-mongodb-mapping/commit/cf7e47befd9d78da648b33bf351c80b3183c509d). It is disabled for the expected SemanticException that the array attribute is not plural path. From HQL perspective, array is still an embeddable field (they require max length in DDL, e.g.), so its attribute is not valid join target.

I simplified the entity to the minimal (so struct has no array field). After reading the https://docs.jboss.org/hibernate/orm/6.6/querylanguage carefully, I chose the following HQL:
```
from Movie as m join m.tagByAuthors as tags where :struct in elements(tags)
```

As per the doc, the joining target could be even embeddable attribute:

<img width="806" alt="Screenshot 2024-09-13 at 17 15 27" src="https://github.com/user-attachments/assets/adefb5e6-5493-41d6-a022-7c4b187b2ab5">

but now I even doubt that is possible.
